src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,120 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+import java.util.Objects;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private T[] elementData;
+    private int size = 0;
+
+    public ArrayList() {
+        elementData = (T[]) new Object[DEFAULT_CAPACITY];
+    }
+
     @Override
     public void add(T value) {
-
+        if (size == elementData.length) {
+            grow();
+        }
+        elementData[size++] = value;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index == size) {
+            add(value);
+        } else if (index < 0 || index > size) {
+            throw new ArrayListIndexOutOfBoundsException("Index not valid");
+        } else {
+            if (size == elementData.length) {
+                grow();
+            }
+            System.arraycopy(elementData, index, elementData, index + 1, size - index);
+            elementData[index] = value;
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
+        if (checkIndex(index)) {
+            return elementData[index];
+        }
         return null;
     }
 
     @Override
     public void set(T value, int index) {
-
+        if (checkIndex(index)) {
+            elementData[index] = value;
+        }
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        T value = null;
+        if (checkIndex(index)) {
+            value = elementData[index];
+            System.arraycopy(elementData, index + 1, elementData, index, size - index - 1);
+            elementData[--size] = null;
+        }
+        return value;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        T value = null;
+        int index = -1;
+        for (int i = 0; i < size; i++) {
+            if (Objects.equals(element, elementData[i])) {
+                index = i;
+                break;
+            }
+        }
+        value = checkRemoveIndex(index);
```

Comment:
If you implement above comment, will we need checkRemoveIndex() method at all or we will definitely know that after cycle is passed we do not have proper object to delete?
And will we need value and index variables?